src/sentry/scim/endpoints/utils.py

```diff
@@ -0,0 +1,162 @@
+from rest_framework.response import Response
+
+from sentry.api.bases.organization import OrganizationEndpoint, OrganizationPermission
+from sentry.api.paginator import BadPaginationError, GenericOffsetPaginator
+from sentry.models import AuthProvider
+from sentry.utils.cursors import Cursor
+
+from .constants import SCIM_API_LIST
+
+SCIM_CONTENT_TYPES = ["application/json", "application/json+scim"]
+
+
+class SCIMClientNegotiation(BaseContentNegotiation):
+    # SCIM uses the content type "application/json+scim"
+    # which is just json for our purposes.
+    def select_parser(self, request, parsers):
+        """
+        Select the first parser in the `.parser_classes` list.
+        """
+        for parser in parsers:
+            if parser.media_type in SCIM_CONTENT_TYPES:
+                return parser
+
+    def select_renderer(self, request, renderers, format_suffix):
+        """
+        Select the first renderer in the `.renderer_classes` list.
+        """
+        for renderer in renderers:
+            if renderer.media_type in SCIM_CONTENT_TYPES:
+                return (renderer, renderer.media_type)
+
+
+class OrganizationSCIMPermission(OrganizationPermission):
+    scope_map = {
+        "GET": ["member:read", "member:write", "member:admin"],
+        "POST": ["member:write", "member:admin"],
+        "PATCH": ["member:write", "member:admin"],
+        "PUT": ["member:write", "member:admin"],
+        "DELETE": ["member:admin"],
+    }
+
+    def has_object_permission(self, request, view, organization):
+        result = super().has_object_permission(request, view, organization)
+        # The scim endpoints should only be used in conjunction with a SAML2 integration
+        if not result:
+            return result
+        try:
+            auth_provider = AuthProvider.objects.get(organization=organization)
+        except AuthProvider.DoesNotExist:
+            return False
+        if not auth_provider.flags.scim_enabled:
+            return False
+
+        return True
+
+
+class SCIMEndpoint(OrganizationEndpoint):
+    permission_classes = (OrganizationSCIMPermission,)
+    content_negotiation_class = SCIMClientNegotiation
+
+    def paginate(
+        self,
+        request,
+        on_results=None,
+        paginator=None,
+        paginator_cls=GenericOffsetPaginator,
+        default_per_page=100,
+        max_per_page=100,
+        cursor_cls=Cursor,
+        **paginator_kwargs,
+    ):
+        assert (paginator and not paginator_kwargs) or (paginator_cls and paginator_kwargs)
+
+        per_page = self.get_per_page(request, default_per_page, max_per_page)
+
+        input_cursor = None
+        if request.GET.get("startIndex"):
+            # XXX: SCIM startIndex param is 1 indexed
+            try:
+                input_cursor = Cursor(0, int(request.GET.get("startIndex")) - 1, 0)
+            except ValueError:
+                raise ParseError(detail="Invalid cursor parameter.")
```

Comment:
I'm not comfortable with this amount of code duplication.

The `Cursor` object is the only part that differs, right? I would extract that section from the original `Endpoint.paginate` method into an instance method, and then override the instance method here.

Reply:
good call on this. I was able to abstract all custom logic out into params passed in along with some help from how functional the base class is written. only addition to parent class is a variable called "cursor_name" that we override to "startIndex" in our child class. check out the latest commit for the diff.

Reply:
Great!